Add compilation and installation steps

elomscansio opened the pull request:
I haven't yet understood how to compile and install nmstate from source

liangwen12year commented on CONTRIBUTING.md:
````diff
@@ -164,3 +176,36 @@ Do your best to follow the clean code guidelines.
+
+### Compilation 
+```
+make
````
I think that we need to guarantee that the system installed the official compiler for the rust programming language first, here is the link https://rustup.rs/.

elomscansio replied:
I have added it

ffmancera replied:
@liangwen12year this comment is confusing. Instead of referring them to rust up, the right way to check everything is fine is running `cargo --version`.

liangwen12year replied:
Sorry, rust up is specifically for the experienced developer, @elomscansio , please ignore my comment.

liangwen12year commented on CONTRIBUTING.md:
```diff
@@ -24,6 +24,18 @@ serving the tests of Nmstate.
+
+- `./rust/src/lib/ovsdb/` Contains the code related to ovsdb communication and structures. 
+
+- `./rust/src/cli/` - Contains command lines tools.
```
There are other subdirs under the `./rust/src/lib` like `policy`, `query_apply`, `statistic` etc. It might be good to also document them here.

elomscansio commented:
I have made the necessary changes

ffmancera commented on CONTRIBUTING.md:
````diff
@@ -164,3 +187,39 @@ Do your best to follow the clean code guidelines.
+
+### Compilation 
+```
+#Check that the system installed the official compiler for the rust programming language
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
````
Instead of this, could we ask the users to do `cargo --version`?

liangwen12year replied:
Sure, if the system does not have the `cargo` installed, then install `cargo` from the package manager.

ffmancera commented on CONTRIBUTING.md:
````diff
@@ -164,3 +187,39 @@ Do your best to follow the clean code guidelines.
+sudo make install
+```
+
+__Note__: In order to report current state, installing Nmstate as a non-root user should be enough. For change support, install the package as root.
````
I do not think this comment is correct. Installing in most situation must be do a root. I think you are mixing executing with installing.

A better note could be: `In order to report current state, running nmstatectl as a non-root user should be enough. For apply support, run nmstatectl as root.`

ffmancera reviewed (changes requested):
Please, drop the commit related to the code structure documentation. It is confusing and not related to the PR.

elomscansio committed:
Update code structure section in CONTRIBUTING.md (#2592)

The CONTRIBUTING.md file had outdated information regarding the code structure,
which has now been updated to accurately reflect the current structure.

Signed-off-by: Emmanuel Elom <elomemmanuel007@gmail.com>

elomscansio commented:
i've made changes

ihuguet commented on CONTRIBUTING.md:
```diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+
+### Dependencies
+
+Nmstate requires NetworkManager 1.26 or later version installed and started.
+
+In order to support specific capabilities, additional packages are required:
+
+OpenVSwitch support through the NM provider requires NetworkManager-ovs.
+
+Extended ovsdb support requires openvswitch-2.11 and python3-openvswitch2.11 or greater.
+
+Team interface support through the NM provider requires NetworkManager-team
```
These are not dependencies to build, they are dependencies to run, no need to explain this here (they are also a bit outdated).

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+
+#### Install build tool
+```
+sudo dnf install cargo # Fedora
+sudo yum install cargo # RHEL
+cargo --version
````
RHEL uses dnf now, so no need to have 2 different lines for them.
Also, there are other distributions like Ubuntu/Debian that the developer might be using. Better to say something more generic like "install cargo from your distribution's repositories".
And finally, for development is normally prefered to install cargo using "rustup" (search the official documentation to see how). So better to indicate installing with rustup, and distribution's repositories as an alternative.

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+### Compilation 
+```
+cd rust
+cargo build --all # Debug version
+cargo build --all --release # Release version
````
No need to put `--all`.

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+
+___Note:___ Installing from source is optional. After compilation, the executable can be run directly.
+```
+target/release/nmstatectl show # Debug compilation
+target/debug/nmstatectl show # Release compilation
````
The comments are wrong.

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+### Installation (Optional)
+```
+# From the root dir
+sudo dnf install cargo make python3 python3-toolz #Fedora
+sudo yum install cargo make python3 python3-toolz #RHEL
````
Again, better not to mention specific distros, just say the packages to install. What is python3-toolz? I think you need python3-setuptool, not toolz.

elomscansio replied:
Thanks

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+cargo build --all --release # Release version
+```
+
+___Note:___ Installing from source is optional. After compilation, the executable can be run directly.
````
Please change `__Note:__` to `Note:`. This fille will be read as plain text, normally.

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,57 @@ Do your best to follow the clean code guidelines.
+nmstatectl show
+```
+
+__Note__: In order to report current state, running nmstatectl as a non-root user should be enough. For apply support, run nmstatectl as root.
````
Same, only `Note:`

elomscansio committed:
Add compilation and installation steps

This commit adds detailed steps for compiling and installing the project

Resolves: #2594

Signed-off-by: Emmanuel Elom <elomemmanuel007@gmail.com>

elomscansio commented:
I have made the requested changes

liangwen12year commented on CONTRIBUTING.md:
````diff
@@ -184,52 +184,55 @@ Ref: Book: Clean Code by Robert C. Martin (Uncle Bob)
+```bash
+# Install Cargo using rustup (recommended for development)
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
+source $HOME/.cargo/env
````
@elomscansio , sorry for confusing you, rust up is specifically for the experienced developer,  you can remove the part above, we can ask the users to do `cargo --version`,  if the system does not have the cargo installed, then install cargo from the package manager of the distribution.

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,60 @@ Do your best to follow the clean code guidelines.
+```
+
+#### Install build tool
+```bash
+# Install Cargo using rustup (recommended for development)
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
+source $HOME/.cargo/env
+cargo --version
+
+# Alternatively, install Cargo from your distribution's repositories
+sudo dnf install cargo  # Fedora, RHEL (using dnf)
+sudo yum install cargo  # Legacy RHEL (using yum)
+sudo apt-get install cargo  # Ubuntu, Debian
+```
+Note:
+  - Rust developers are encouraged to install Cargo using rustup, which is recommended for development purposes.
+  - Instructions for installing Cargo from distribution repositories are provided as an alternative for those who prefer that method.
+  - Distribution-specific commands for installing Cargo are included for Fedora (using dnf), Legacy RHEL (using yum), Ubuntu, and Debian.
````
This is a very long explanation for something that it's very simple. Also, I insist in that we don't need to provide the exact command to install from every distribution. I would just put something like:

> Install cargo using rustup (recommended): URL to official website
> Alternatively, you can install cargo from your distribution's repositories.

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,60 @@ Do your best to follow the clean code guidelines.
+For installing `cargo` refer to [Install build tool](#install-build-tool)
+```bash
+# Install dependencies using package manager
+sudo dnf install make python3 python3-setuptool  # Fedora, RHEL (using dnf)
+sudo yum install make python3 python3-setuptool  # Legacy RHEL (using yum)
+sudo apt-get install make python3  # Ubuntu, Debian
````
Same, just list the packages without providing specific commands.

elomscansio commented:
I have made the requested changes

ihuguet commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,43 @@ Do your best to follow the clean code guidelines.
+```
+
+#### Install build tool
+Install cargo using rustup (recommended): https://sh.rustup.rs.
````
This link is not to the installation instructions. Better link to the instructions.

ihuguet commented on CONTRIBUTING.md:
```diff
@@ -179,3 +179,43 @@ Do your best to follow the clean code guidelines.
 
 Ref: Book: Clean Code by Robert C. Martin (Uncle Bob)
+
+## Compilation and Installation
```
Compilation and installation from source

ihuguet reviewed (changes requested):
I see it better now. However, I still see it too extense, with many subchapters with small contents on each one. Do you think you can put everything in a single chapter and in a bit more concise way?

elomscansio commented:
I have made the changes

cathay4t commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,29 @@ Do your best to follow the clean code guidelines.
+To install nmstate from source, install `cargo`, `make`, `python3`, and `python3-setuptools` from your distribution's repositories. It is recommended to use `rustup` for installing `cargo`: https://www.rust-lang.org/tools/install.
+
+```bash
+sudo make install # To compile and install the release version
````
`sudo PREFIX=/usr make install`

cathay4t commented on CONTRIBUTING.md:
````diff
@@ -179,3 +179,29 @@ Do your best to follow the clean code guidelines.
+
+```bash
+cd rust
+cargo build # Build the debug version add --release flag to build the release version
````
Please use `make` from the top level of code

ihuguet commented:
Mostly looks good to me. Just change the title to "Compilation and installation from source".

elomscansio commented:
> Mostly looks good to me. Just change the title to "Compilation and installation from source".

done

ihuguet reviewed (approved):
Good work!
